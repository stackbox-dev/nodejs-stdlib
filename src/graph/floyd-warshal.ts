import { Edge, TransitiveClosure, INF, add } from "./interfaces";
import { getNodesFromEdges } from "./utils";

export function buildDAGTransitiveClosure<T>(
  edges: Edge<T>[],
): TransitiveClosure<T>[] {
  const nodes = getNodesFromEdges(edges);
  const N = nodes.length;
  if (N > INF / 2) {
    throw new Error("too many nodes " + N);
  }
  const nodeIndex = new Map(nodes.map((n, i) => [n, i]));
  const indexedEdges = edges.map((e) => {
    const from = nodeIndex.get(e[0])!;
    const to = nodeIndex.get(e[1])!;
    return [from, to, e[2]] as [number, number, number];
  });

  const dm = buildDMUsingFloydWarshal(nodes.length, indexedEdges);
  return buildClosureFromDM(nodes, dm);
}

// build Distance Matrix using Floyd Warshal
// https://en.wikipedia.org/wiki/Floyd%E2%80%93Warshall_algorithm
// N - number of nodes/vertexes, edges - array of [from, to, distance]
export function buildDMUsingFloydWarshal(
  N: number,
  edges: [number, number, number][],
): number[][] {
  const dm: number[][] = [];
  for (let i = 0; i < N; i++) {
    const row: number[] = [];
    dm[i] = row;
    for (let j = 0; j < N; j++) {
      row[j] = INF;
    }
  }

  for (const [from, to, distance] of edges) {
    dm[from][to] = distance;
  }
  for (let i = 0; i < N; i++) {
    dm[i][i] = 0;
  }

  for (let k = 0; k < N; k++) {
    for (let from = 0; from < N; from++) {
      for (let to = 0; to < N; to++) {
        const oldDist = dm[from][to];
        const newDist = add(dm[from][k], dm[k][to]);
        if (newDist < oldDist) {
          dm[from][to] = newDist;
        }
      }
    }
  }
  return dm;
}

export function buildClosureFromDM<T>(nodes: T[], dm: number[][]) {
  const closure: TransitiveClosure<T>[] = [];
  const N = nodes.length;
  for (let from = 0; from < N; from++) {
    for (let to = 0; to < N; to++) {
      const distance = dm[from][to];
      if (distance === INF) {
        continue;
      }
      closure.push({ child: nodes[to], parent: nodes[from], distance });
    }
  }
  return closure;
}
