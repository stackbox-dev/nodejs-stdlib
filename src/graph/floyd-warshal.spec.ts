import { buildDAGTransitiveClosure } from "./floyd-warshal";

const t1: [string, string][] = [
  [
    "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
    "57fea751-6c21-4eb1-9262-ea97ff45553c",
  ],
  [
    "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
    "f03626f2-1852-45b8-b6ad-b59bde38b3fd",
  ],
  [
    "9c550989-3e19-4cf5-ab28-b8e0881e2dac",
    "5c1a9dca-f58a-4f49-9909-576c48ab20c1",
  ],
  [
    "57fea751-6c21-4eb1-9262-ea97ff45553c",
    "322c07a4-06ce-4766-964c-2ccc7d977cbc",
  ],
];

test("success topo - t1", () => {
  expect(buildDAGTransitiveClosure(t1)).toMatchSnapshot();
});

const t2: [string, string][] = [
  [
    "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
    "57fea751-6c21-4eb1-9262-ea97ff45553c",
  ],
  [
    "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
    "f03626f2-1852-45b8-b6ad-b59bde38b3fd",
  ],
  [
    "57fea751-6c21-4eb1-9262-ea97ff45553c",
    "322c07a4-06ce-4766-964c-2ccc7d977cbc",
  ],
];

test("success topo - t2", () => {
  expect(buildDAGTransitiveClosure(t2)).toMatchSnapshot();
});
