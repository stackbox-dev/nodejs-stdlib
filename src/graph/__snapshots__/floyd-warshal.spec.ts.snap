// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success topo - t1 1`] = `
Array [
  Object {
    "child": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
    "distance": 0,
    "parent": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
  },
  Object {
    "child": "57fea751-6c21-4eb1-9262-ea97ff45553c",
    "distance": 1,
    "parent": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
  },
  Object {
    "child": "f03626f2-1852-45b8-b6ad-b59bde38b3fd",
    "distance": 1,
    "parent": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
  },
  Object {
    "child": "322c07a4-06ce-4766-964c-2ccc7d977cbc",
    "distance": 2,
    "parent": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
  },
  Object {
    "child": "9c550989-3e19-4cf5-ab28-b8e0881e2dac",
    "distance": 0,
    "parent": "9c550989-3e19-4cf5-ab28-b8e0881e2dac",
  },
  Object {
    "child": "5c1a9dca-f58a-4f49-9909-576c48ab20c1",
    "distance": 1,
    "parent": "9c550989-3e19-4cf5-ab28-b8e0881e2dac",
  },
  Object {
    "child": "57fea751-6c21-4eb1-9262-ea97ff45553c",
    "distance": 0,
    "parent": "57fea751-6c21-4eb1-9262-ea97ff45553c",
  },
  Object {
    "child": "322c07a4-06ce-4766-964c-2ccc7d977cbc",
    "distance": 1,
    "parent": "57fea751-6c21-4eb1-9262-ea97ff45553c",
  },
  Object {
    "child": "f03626f2-1852-45b8-b6ad-b59bde38b3fd",
    "distance": 0,
    "parent": "f03626f2-1852-45b8-b6ad-b59bde38b3fd",
  },
  Object {
    "child": "5c1a9dca-f58a-4f49-9909-576c48ab20c1",
    "distance": 0,
    "parent": "5c1a9dca-f58a-4f49-9909-576c48ab20c1",
  },
  Object {
    "child": "322c07a4-06ce-4766-964c-2ccc7d977cbc",
    "distance": 0,
    "parent": "322c07a4-06ce-4766-964c-2ccc7d977cbc",
  },
]
`;

exports[`success topo - t2 1`] = `
Array [
  Object {
    "child": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
    "distance": 0,
    "parent": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
  },
  Object {
    "child": "57fea751-6c21-4eb1-9262-ea97ff45553c",
    "distance": 1,
    "parent": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
  },
  Object {
    "child": "f03626f2-1852-45b8-b6ad-b59bde38b3fd",
    "distance": 1,
    "parent": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
  },
  Object {
    "child": "322c07a4-06ce-4766-964c-2ccc7d977cbc",
    "distance": 2,
    "parent": "5ab489c9-824c-4f2b-a021-54dedd2db6bd",
  },
  Object {
    "child": "57fea751-6c21-4eb1-9262-ea97ff45553c",
    "distance": 0,
    "parent": "57fea751-6c21-4eb1-9262-ea97ff45553c",
  },
  Object {
    "child": "322c07a4-06ce-4766-964c-2ccc7d977cbc",
    "distance": 1,
    "parent": "57fea751-6c21-4eb1-9262-ea97ff45553c",
  },
  Object {
    "child": "f03626f2-1852-45b8-b6ad-b59bde38b3fd",
    "distance": 0,
    "parent": "f03626f2-1852-45b8-b6ad-b59bde38b3fd",
  },
  Object {
    "child": "322c07a4-06ce-4766-964c-2ccc7d977cbc",
    "distance": 0,
    "parent": "322c07a4-06ce-4766-964c-2ccc7d977cbc",
  },
]
`;
